klib/syslog.c

```diff
@@ -0,0 +1,338 @@
+    boolean (*attach_console_driver)(struct console_driver *driver) =
+            get_sym("attach_console_driver");
+    if (!get_root_tuple || !intern || !get || !runtime_memcmp || !fsfile_get_length ||
+            !get_kernel_heaps || !udp_new || !attach_console_driver ||
+            !(kfuncs.fsfile_open_or_create = get_sym("fsfile_open_or_create")) ||
+            !(kfuncs.fsfile_get_writer = get_sym("fsfile_get_writer")) ||
+            !(kfuncs.allocate_sg_list = get_sym("allocate_sg_list")) ||
+            !(kfuncs.sg_list_tail_add = get_sym("sg_list_tail_add")) ||
+            !(kfuncs.deallocate_sg_list = get_sym("deallocate_sg_list")) ||
+            !(kfuncs.runtime_memcpy = get_sym("runtime_memcpy")) ||
+            !(kfuncs.register_timer = get_sym("kern_register_timer")) ||
+            !(kfuncs.timm_dealloc = get_sym("timm_dealloc")) ||
+            !(kfuncs.fs_rename = get_sym("fs_rename")) ||
+            !(kfuncs.dns_gethostbyname = get_sym("dns_gethostbyname")) ||
+            !(kfuncs.netif_get_default = get_sym("netif_get_default")) ||
+            !(kfuncs.ipaddr_ntoa_r = get_sym("ipaddr_ntoa_r")) ||
+            !(kfuncs.pbuf_alloc = get_sym("pbuf_alloc")) ||
+            !(kfuncs.pbuf_free = get_sym("pbuf_free")) ||
+            !(kfuncs.now = get_sym("now")) ||
+            !(kfuncs.gmtime_r = get_sym("gmtime_r")) ||
+            !(kfuncs.rsnprintf = get_sym("rsnprintf")) ||
+            !(kfuncs.udp_sendto = get_sym("udp_sendto"))) {
+        rprintf("syslog: kernel symbols not found\n");
+        return KLIB_INIT_FAILED;
+    }
+    tuple root = get_root_tuple();
+    buffer output = get(root, sym(syslog_output));
+    if (!output) {
+        rprintf("syslog output not specified\n");
+        return KLIB_INIT_FAILED;
+    } else if (!is_string(output)) {
+        rprintf("invalid syslog output\n");
+        return KLIB_INIT_FAILED;
+    }
+    syslog.h = heap_general(get_kernel_heaps());
+    if ((buffer_length(output) == 4) && !runtime_memcmp(buffer_ref(output, 0), "file", 4)) {
+        syslog.output_file = get(root, sym(syslog_file));
+        if (!syslog.output_file) {
+            rprintf("syslog file not specified\n");
+            return KLIB_INIT_FAILED;
+        } else if (!is_string(syslog.output_file)) {
+            rprintf("invalid syslog file\n");
+            return KLIB_INIT_FAILED;
+        }
+        fsfile file = kfuncs.fsfile_open_or_create(syslog.output_file);
+        if (!file) {
+            rprintf("cannot create syslog output file\n");
+            return KLIB_INIT_FAILED;
+        }
+        syslog.fs_write = kfuncs.fsfile_get_writer(file);
+        syslog.file_offset = fsfile_get_length(file);   /* append to existing contents */
+        syslog.file_flush_timer = INVALID_ADDRESS;
+        init_closure(&syslog.file_flush, syslog_timer_func);
+        syslog.driver.write = syslog_file_write;
+    } else if ((buffer_length(output) == 3) && !runtime_memcmp(buffer_ref(output, 0), "udp", 3)) {
```

Comment:
Is there any particular reason syslog output has to be either file or udp but not both?

Reply:
I changed the code so that file and udp output ca be enabled at the same time. Now, the `syslog_output` manifest option takes as value an array, e.g. `[udp]` to enable udp output and `[file udp]` to enable both file and udp output.

Reply:
Is syslog_output even necessary? The presence of either syslog_file or syslog_server should be sufficient to indicate which outputs are being used.

Reply:
Don't mean to be a nuisance, but maybe it would be preferable to have just a "syslog" tuple that has one or more of "file", "server" and "port" attributes.

Reply:
Ok, makes sense, I will change the code to use a single syslog tuple with one or more attributes, among which there can be log rotation parameters

Reply:
Configuration settings for the syslog klib are now grouped in a `syslog` manifest option, as explained in the updated description of this PR.